src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -2991,27 +2991,62 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       fun.body match {
         // we can compare arguments and parameters by name because there cannot be a binder between
         // the function's valdefs and the Apply's arguments
-        case Apply(meth, args) if (vparams corresponds args) { case (p, Ident(name)) => p.name == name case _ => false } =>
-          // We're looking for a method (as indicated by FUNmode in the silent typed below),
-          // so let's make sure our expected type is a MethodType
-          val methArgs = NoSymbol.newSyntheticValueParams(argProtos map { case NoType => WildcardType case tp => tp })
-
-          silent(_.typed(meth, mode.forFunMode, MethodType(methArgs, resProto))).fold(EmptyTree: Tree){ methTyped =>
-            if (context.undetparams.isEmpty) {
-              val numVparams = vparams.length
+        // If all vparams are constrained by the method application, see if we can derive expected types for them.
+        // Note that not all method arguments need be references to a function param.
+        case Apply(meth, args) =>
+          // Map param with missing param type to the argument it's passed as in the eta-expanded method application
+          // This list specifies a way to compute the expected parameter type for each of our function's arguments in order.
+          // Either we already know it, and then we have a Type, or we don't, and then it's an index `idx` into
+          // the arguments passed to `meth`, so we can derive it from its MethodType
+          // (based on where the function's parameter is applied to `meth`)
+          val formalsFromApply =
+            vparams.map { vd =>
+              if (!vd.tpt.isEmpty) Right(vd.tpt.tpe)
+              else Left(args.indexWhere {
+                case Ident(name) => name == vd.name
+                case _           => false // TODO: i think we need to deal with widening conversions too??
+              })
+            }
+
+          // If some of the vparams without type annotation was not applied to `meth`,
+          // we're not going to learn enough from typing `meth` to determine them.
+          if (formalsFromApply.exists{ case Left(-1) => true case _ => false }) EmptyTree
```

Comment:
`.exists(_ == Left(-1))` should work, or you mind the `Left` allocation?

Reply:
True, or even `.contains(Left(-1))`. I was probably experimenting with fancier logic, but in the end it came out simpler.